### DISCUSSION TITLE
How to update the list of inputs and outputs

### DISCUSSION THREAD

**Discussion by FDelporte** [atsushieno/ktmidi#84] [General]
I'm using this to get the list of MIDI inputs and outputs:

```
private val access: MidiAccess = getMidiAccessApi("JVM") 

private fun getMidiAccessApi(api: String?) = when (api) {
    "EMPTY" -> EmptyMidiAccess()
    "JVM" -> JvmMidiAccess()
    else -> JvmMidiAccess()
}
       
access.inputs.iterator().asSequence().toList()
```

When I (dis)connect a device while the app is running, it doesn't get added to this list, and I need to restart the application. How can I "force" the access to refresh it list of devices?

**Comment by atsushieno** [comment]
`JvmMidiAccess.inputs` invokes [`MidiSystem.getMidiDeviceInfo()`](https://docs.oracle.com/javase/8/docs/api/javax/sound/midi/MidiSystem.html#getMidiDeviceInfo--) and [`MidiDevice.getTransmitters()`](https://docs.oracle.com/javase/8/docs/api/javax/sound/midi/MidiDevice.html#getTransmitters--). I believe they both return a cold snapshot list. It invokes `MidiDevice.getTransmitters()` for each `MidiDevice`, and each invocation would happen only when the previous results from `getTransmitter()` are handled, so it would feel like partially cold though.

If you get `JvmMidiAccess.inputs` in another list, that would give you the updated snapshot list at the accessed time.

**Reply by FDelporte** [reply]
Sorry, don't understand exactly what I need to call to refresh the list...

**Reply by atsushieno** [reply]
Oh, okay let me describe in other words: `MidiAccess.inputs` is a Kotlin property, which would be mapped to Java `getInputs()` method. With the code example below:

```
val access = JvmMidiAccess()
val inputs1 = access.inputs // this creates a snapshot of the list when this code is run.
val list1 = access.inputs.iterator().asSequence().toList() // or simply, access.inputs.toList()

// connect your MIDI device to your PC here.
...

val inputs2 = access.inputs // this creates a snapshot of the list when this code is run.
val list2 = access.inputs.iterator().asSequence().toList()
```

list1 "frozen" when `(Sequence<T>).toList()` is invoked (more specifically, when `iterator()` is invoked), where "frozen" means that the list of the iterated devices is fixed at the state when you invoked it.

When you try to get `(Jvm)MidiAccess.inputs` property after you connected your MIDI device, this `getInputs()` calls `MidiSystem.getMidiDeviceInfo()` (in `javax.sound.midi` API), which returns the latest list of devices to you, and our `MidiAccess.inputs` translates the resulting ports (in the devices in the device list).

There is no caching at ktmidi side, so if you still do not see the new device in the updated list, you are likely seeing outdated list (in this case, `list1`). This kind of problem could occur on Jetpack Compose where your code isn't actually run.

That's what I meant by `get JvmMidiAccess.inputs in another list` (`input2` in the example above).

**Comment by FDelporte** [comment]
I rewrote my code to this, to always use a new JvmMidiAccess at each request of the inputs, but I still don't get new connections until the app restarts...

```
    fun getInputs(): List<MidiPortDetails> {
        logger.info(
            "Test input list: {}",
            JvmMidiAccess().inputs.toList().stream().map { i -> i.name }.collect(Collectors.joining(","))
        )
        return JvmMidiAccess().inputs.toList()
    }
```

**Reply by atsushieno** [reply]
It does not matter whether you create a new `JvmMidiAccess` or not - `inputs` from the same `JvmMidiAccess` instance would result in the same. You need to call your `getInputs()` and get a refresh (up to date) list again *after* you connected or disconnected your MIDI device.

I'm sure you will have the same problem if you use `javax.sound.midi` directly.

**Reply by FDelporte** [reply]
Yes, that's what I do... but still get the old list...

**Reply by atsushieno** [reply]
We figure it out onsite (kind of!)

**Comment by atsushieno** [comment]
We figured out that it is rather an upstream javax.sound.midi issue. https://stackoverflow.com/questions/41123620/update-list-of-midi-devices-in-java
